compile/inc/src/main/scala/sbt/inc/MemberRefInvalidator.scala

```diff
@@ -79,12 +79,19 @@ private[inc] class MemberRefInvalidator(log: Logger) {
     "MemberReferenceInvalidator.get should be called when name hashing is enabled " +
       "and in that case we shouldn't have SourceAPIChange as an api change."
 
-  private class InvalidateUnconditionally[T](memberRef: Relation[File, T]) extends (T => Set[File]) {
+  private class InvalidateUnconditionally[T](memberRef: Relation[File, T], apiChange: APIChange[_]) extends (T => Set[File]) {
     def apply(from: T): Set[File] = {
       val invalidated = memberRef.reverse(from)
-      if (invalidated.nonEmpty)
-        log.debug(s"The following member ref dependencies of $from are invalidated:\n" +
-          formatInvalidated(invalidated))
+      if (invalidated.nonEmpty) {
+        apiChange match {
```

Comment:
Isn't it better to introduce `InvalidateDueToMacroDefinition` class and put special logging logic there? This way you can let virtual dispatch do the job instead of introducing a pattern match. Also, this way you pattern match on APIChange just once: in `MemberRefInvalidator.get` method.